### PR TITLE
MD-TRT Support, Compile/Export, C++ and Python 

### PR DESCRIPTION
**Pull request by narendasan**
# Description

Opening this to test the CI 

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: core/runtime/execute_engine.cpp [hunk 1]
```diff
@@ -209,6 +207,27 @@ void create_output_allocator(c10::intrusive_ptr<TRTEngine> compiled_engine) {
 }
 
 std::vector<at::Tensor> execute_engine(std::vector<at::Tensor> inputs, c10::intrusive_ptr<TRTEngine> compiled_engine) {
+  // All inputs are expected to be on CUDA. Warn and move any that are not.
+  for (auto& inp : inputs) {
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp:211]: I would like to remove this but didnt have time to check if the device operations in python suppress this correctly

File: core/runtime/execute_engine.cpp [hunk 2]
```diff
@@ -209,6 +207,27 @@ void create_output_allocator(c10::intrusive_ptr<TRTEngine> compiled_engine) {
...
+      inp = inp.cuda();
+    }
+  }
+
+#ifdef ENABLE_TRT_NCCL_COLLECTIVES
+  // Lazy one-shot NCCL bind: fires on the first real execute_engine call when
+  // the constructor-time bind was deferred (e.g. no collective had been issued
+  // at construction time, or for serialized programs loaded inline where there
+  // is no Python _TorchTensorRTModule.forward wrapper).
+  if (compiled_engine->is_md && !compiled_engine->nccl_initialized) {
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp:226]: Not entirely sure this is necessary

File: core/runtime/register_jit_hooks.cpp [hunk 3]
```diff
@@ -108,6 +108,45 @@ static auto TORCHTRT_UNUSED TRTEngineTSRegistrtion =
...
         .def_property("streamable_device_memory_budget", &TRTEngine::get_streamable_device_memory_budget)
         .def_property("automatic_device_memory_budget", &TRTEngine::get_automatic_device_memory_budget)
+        .def_readonly("is_md", &TRTEngine::is_md)
+        .def_readonly("rank", &TRTEngine::rank)
+        .def_readonly("world_size", &TRTEngine::world_size)
+#ifdef ENABLE_TRT_NCCL_COLLECTIVES
+        .def(
+            "set_group_name",
+            [](c10::intrusive_ptr<TRTEngine> self, std::string group_name) {
+              self->group_name = group_name;
```
Comment (narendasan) [inline core/runtime/register_jit_hooks.cpp]: We should have a test where we switch engines between process groups

File: core/runtime/TRTEngine.cpp [hunk 4]
```diff
@@ -519,6 +561,75 @@ void TRTEngine::set_resource_allocation_strategy(TRTEngine::ResourceAllocationSt
...
 
+#ifdef ENABLE_TRT_NCCL_COLLECTIVES
+bool TRTEngine::bind_nccl_comm() {
+  // When group_name is empty (e.g. engine loaded from a serialized
+  // ExportedProgram where the Python TorchTensorRTModule wrapper was
+  // inlined and set_group_name() was never called), auto-resolve the
+  // process group from the c10d registry.  PyTorch assigns sequential
+  // numeric names ("0", "1", ...) to process groups; probe until we
+  // find one with an NCCL backend.
+  if (this->group_name.empty() && this->is_md) {
```
Comment (narendasan) [inline core/runtime/TRTEngine.cpp:564]: We should only do this if there is one available group. If there are multiple NCCL groups available we should tell the user to manually select

File: docsrc/tutorials/deployment/distributed_inference.rst [hunk 5]
```diff
@@ -74,67 +85,449 @@ handles DTensor inputs correctly:
...
+    # Manual row-parallel wrapper (replaces DTensor RowwiseParallel)
+    class RowParallelLinear(torch.nn.Module):
+        def __init__(self, linear, group_name):
+            super().__init__()
+            self.linear = linear
+            self.group_name = group_name
+
+        def forward(self, x):
+            out = self.linear(x)
+            out = torch.ops._c10d_functional.all_reduce(out, "sum", self.group_name)
```
Comment (narendasan) [inline docsrc/tutorials/deployment/distributed_inference.rst:196]: Lets dig into this more after the PR lands

File: py/torch_tensorrt/distributed/run/__init__.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+import os
+import signal
+import subprocess
+import sys
+from typing import Optional
+
+logger = logging.getLogger("torchtrtrun")
+
+
+def _get_nccl_lib_dir() -> Optional[str]:
```
Comment (narendasan) [inline py/torch_tensorrt/distributed/run/__init__.py]: Move into its own file

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 7]
```diff
@@ -229,6 +230,9 @@ def __init__(
         if self.serialized_engine is not None and not self.settings.lazy_engine_init:
             self.setup_engine()
 
+        self._nccl_comm: Optional[Any] = None
+        self._has_nccl_ops: bool = False
+
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: this should be set before the `self.setup_engine()`

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 8]
```diff
@@ -291,6 +351,18 @@ def setup_engine(self) -> None:
         if self.settings.enable_weight_streaming:
             self.set_default_device_memory_budget()
         self.context = self.engine.create_execution_context()
+
+        # Detect NCCL collective layers via engine inspector
+        inspector = self.engine.create_engine_inspector()
+        engine_json = inspector.get_engine_information(trt.LayerInformationFormat.JSON)
+        self._has_nccl_ops = "NCCL" in engine_json or "AllReduce" in engine_json
+
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: something like this works
engine_json_lower = engine_json.lower()
self._has_nccl_ops = "dist_collective" in engine_json_lower or "nccl" in engine_json_lower or "allreduce" in engine_json_lower

File: examples/distributed_inference/tensor_parallel_simple_example.py
Comment (apbose) [inline examples/distributed_inference/tensor_parallel_simple_example.py:1]: with the present changes, this example wont run with just torchtrtrun since that uses torchrun. Is the expectation that the user do LD_LIBRARY_PATH and run the test?

Comment (apbose) [reply examples/distributed_inference/tensor_parallel_simple_example.py:1]: Or accordingly we need to change the initialization code in this

File: py/torch_tensorrt/distributed/_distributed.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+            ):
+                seen.add(id(attr_val))
+                attr_val.set_group_name(group_name)
+
+    # Also pin on any engines in the thread-local registry (handles torch.compile).
+    for engine in getattr(_state, "md_engines", None) or []:
+        if id(engine) not in seen:
+            seen.add(id(engine))
+            if getattr(engine, "is_md", False):
+                engine.set_group_name(group_name)
```
Comment (apbose) [inline py/torch_tensorrt/distributed/_distributed.py:207]: this seems to be designed in a way that it works for only C++ runtime. Why dont we have it for python runtime? Something like maintaining a group_name and nccl_comm instead of getting the active group and binding it once

File: py/torch_tensorrt/distributed/_distributed.py [hunk 10]
```diff
@@ -159,3 +197,10 @@ def set_distributed_group(module: nn.Module, group: Any) -> None:
             ):
                 seen.add(id(attr_val))
                 attr_val.set_group_name(group_name)
+
+    # Also pin on any engines in the thread-local registry (handles torch.compile).
+    for engine in getattr(_state, "md_engines", None) or []:
+        if id(engine) not in seen:
+            seen.add(id(engine))
+            if getattr(engine, "is_md", False):
+                engine.set_group_name(group_name)
```
Comment (apbose) [inline py/torch_tensorrt/distributed/_distributed.py:207]: why is this added?

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 11]
```diff
@@ -309,7 +309,7 @@ def setup_nccl_comm(self) -> None:
         if pg is None or dist.get_backend(pg) != "nccl":
             raise RuntimeError(
                 "Active ProcessGroup must use NCCL backend. "
-                "Use torch_tensorrt.distributed_group(group) to select a non-default group."
+                "Use torch_tensorrt.distributed.distributed_context(group) to select a non-default group."
             )
 
         backend = pg._get_backend(torch.device("cuda"))
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:314]: shouldn't we have a equivalent release_nccl_comm for python runtime as well?

File: py/torch_tensorrt/distributed/_distributed.py (hunk 9 above)
Comment (narendasan) [reply py/torch_tensorrt/distributed/_distributed.py:207]: It doesnt really matter for the python version and the python version is going to change

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py (hunk 11 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:314]: Its not necessary

File: py/torch_tensorrt/distributed/_distributed.py (hunk 10 above)
Comment (narendasan) [reply py/torch_tensorrt/distributed/_distributed.py:207]: We need to track instances for clean up

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py (hunk 11 above)
Comment (apbose) [reply py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:314]: why?

Comment (narendasan) [reply py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:314]: When everything is handled by python, the cleanup seems to get ordered properly since i guess the execution context object is handled by python. We also will be removing this as well with @cehongwang's PR so I am not spending time thinking too much about it as long as it works well enough to pass tests and land in main
